kclvm/tools/src/LSP/src/rename.rs

```diff
@@ -0,0 +1,302 @@
+                                    .filter(|&&loc| {
+                                        // 4.2 filter out the words and remain those whose definition is the target def
+                                        let p = loc.uri.path();
+                                        if let Ok((_, _, _, gs)) = parse_param_and_compile(
+                                            Param {
+                                                file: p.to_string(),
+                                            },
+                                            None,
+                                        ) {
+                                            let kcl_pos = kcl_pos(p, loc.range.start);
+                                            if let Some(symbol_ref) =
+                                                find_def_with_gs(&kcl_pos, &gs, true)
+                                            {
+                                                if let Some(real_def) =
+                                                    gs.get_symbols().get_symbol(symbol_ref)
+                                                {
+                                                    return real_def.get_definition() == def;
+                                                }
+                                            }
+                                        }
+                                        false
+                                    })
+                                    .cloned()
+                                    .collect::<Vec<&Location>>();
+                                // 5. refs to rename actions
+                                let changes =
+                                    refs.into_iter().fold(HashMap::new(), |mut map, location| {
+                                        let uri = &location.uri;
+                                        map.entry(uri.clone()).or_insert_with(Vec::new).push(
+                                            TextEdit {
+                                                range: location.range,
+                                                new_text: new_name.clone(),
+                                            },
+                                        );
+                                        map
+                                    });
+                                return Ok(changes);
+                            }
+                        }
+                    }
+                }
+            }
+            Result::Err("rename failed".to_string())
+        }
+        Result::Err(err) => {
+            return Result::Err(format!(
+                "can not parse symbol path {}, {}",
+                symbol_path, err
+            ));
+        }
+    }
+}
+
+/// Select a symbol by the symbol path
+/// The symbol path should be in the format of: `pkg.sub_pkg:name.sub_name`
+pub fn select_symbol(selector: &ast::SymbolSelectorSpec) -> Option<(String, Option<SymbolRef>)> {
```

Comment:
Here is a more scalable and concise implementation.
```
/// Select a symbol by the symbol path
/// The symbol path should be in the format of: `pkg.sub_pkg:name.sub_name`
pub fn select_symbol(selector: &ast::SymbolSelectorSpec) -> Option<(String, Range)> {
    let mut pkg = PathBuf::from(&selector.pkg_root);
    let pkg_names = selector.pkgpath.split(".");
    for n in pkg_names {
        pkg = pkg.join(n)
    }

    let fields: Vec<&str> = selector.field_path.split(".").collect();
    match pkg.as_path().to_str() {
        Some(pkgpath) => {
            // resolve pkgpath and get the symbol data by the fully qualified name
            if let Ok((prog, _, _, gs)) = parse_param_and_compile(
                Param {
                    file: pkgpath.to_string(),
                },
                None,
            ) {
                if let Some(symbol_ref) = gs
                    .get_symbols()
                    .get_symbol_by_fully_qualified_name(&prog.main)
                {
                    let mut owner_ref = symbol_ref;
                    let mut target = None;
                    for field in fields {
                        let owner = gs.get_symbols().get_symbol(owner_ref).unwrap();
                        target = owner.get_attribute(field, gs.get_symbols(), None);
                        if let Some(target) = target {
                            owner_ref = target;
                        }
                    }

                    let target_symbol = gs.get_symbols().get_symbol(target?)?;
                    return Some((target_symbol.get_name(), target_symbol.get_range()));
                }
            }
            None
        }
        None => None,
    }
}
```

Reply:
Any internal references (such as SymbolRef or ScopeRef) should not be returned after the GlobalState is destroyed. If you just want to quickly locate a symbol, please return the range of this symbol.

Reply:
thx, that's clear. fixed with get_attribute and range. ptal